lib/Backend/arm64/LowerMD.cpp

```diff
@@ -791,64 +791,47 @@ LowererMD::GenerateStackProbe(IR::Instr *insertInstr, bool afterProlog)
-            this->CreateAssign(scratchOpnd, IR::AddrOpnd::New(pLimit, IR::AddrOpndKindDynamicMisc, this->m_func), insertInstr);
-            this->CreateAssign(scratchOpnd, IR::IndirOpnd::New(scratchOpnd, 0, TyMachReg, this->m_func), insertInstr);
-        }
-
-        if (IS_CONST_00000FFF(frameSize))
-        {
-            // If the frame size is small enough, just add the constant.
-            // Does this ever happen with the size of the MinStackJIT constant?
-            instr = IR::Instr::New(Js::OpCode::ADD, scratchOpnd, scratchOpnd,
-                                   IR::IntConstOpnd::New(frameSize, TyMachReg, this->m_func), this->m_func);
-            insertInstr->InsertBefore(instr);
-        }
-        else
-        {
-            // We need a second scratch reg.
-            // If we're probing after the prolog, the reg has already been saved and will be restored.
-            // If not, push and pop it here, knowing that we'll never throw while the stack is whacked.
-            Assert(!afterProlog || this->m_func->m_unwindInfo.GetSavedScratchReg());
-
-            BVUnit scratchBit;
-            IR::Opnd *opnd;
-            if (!afterProlog)
-            {
-                opnd = IR::IndirOpnd::New(IR::RegOpnd::New(nullptr, RegSP, TyMachReg, this->m_func), (int32)0, TyMachReg, this->m_func);
-                instr = IR::Instr::New(Js::OpCode::STR, opnd, this->m_func);
-                scratchBit.Set(RegEncode[SP_ALLOC_SCRATCH_REG]);
-                opnd = IR::RegBVOpnd::New(scratchBit, TyMachReg, this->m_func);
-                instr->SetSrc1(opnd);
-                insertInstr->InsertBefore(instr);
+        // LDIMM r17, &ThreadContext::scriptStackLimitForCurrentThread
+        intptr_t pLimit = m_func->GetThreadContextInfo()->GetThreadStackLimitAddr();
+        this->CreateAssign(scratchOpnd, IR::AddrOpnd::New(pLimit, IR::AddrOpndKindDynamicMisc, this->m_func), insertInstr);
 
-                IR::RegOpnd* regSpOpnd = IR::RegOpnd::New(nullptr, RegSP, TyMachReg, this->m_func);
-                IR::Instr * instrSub = IR::Instr::New(Js::OpCode::SUB, regSpOpnd, regSpOpnd, IR::IntConstOpnd::New(MachRegInt, TyMachReg, this->m_func), this->m_func);
-                insertInstr->InsertBefore(instrSub);
+        // LDR   r17, [r17, #0]
+        this->CreateAssign(scratchOpnd, IR::IndirOpnd::New(scratchOpnd, 0, TyMachReg, this->m_func), insertInstr);
 
-            }
+        AssertMsg(!IS_CONST_00000FFF(frameSize), "For small size we can just add frameSize to r17");
 
-            IR::Opnd *scratchOpnd2 = IR::RegOpnd::New(nullptr, SP_ALLOC_SCRATCH_REG, TyMachReg, this->m_func);
-            this->CreateAssign(scratchOpnd2, IR::IntConstOpnd::New(frameSize, TyMachReg, this->m_func), insertInstr);
+        IR::Opnd *spAllocRegOpnd = nullptr;
+        IR::Opnd *unusedRegForAlignOpnd = nullptr;
+        if (!afterProlog)
+        {
+            // STP r15,r14,[sp, #-16]
```

Comment:
This very unusual. First, x14, x15 are volatile registers, no need to save them. Second, if I'm reading this right, it looks like the saves are done in the function body. You are not allowed to make changes to SP in the middle of a function, unless you are saving SP in the prolog to a reserved nonvolatile register.

Reply:
SP is being changed only when afterProlog is false, so this is in prolog. 
You are right we don't need to save x14, x15 so I removed this code

---
In reply to: [144412145](https://github.com/Microsoft/ChakraCore/pull/3943#discussion_r144412145) [](ancestors = 144412145)